jsign-core/src/main/java/net/jsign/jca/IMDS2Client.java

```diff
@@ -0,0 +1,146 @@
+
+/**
+ * Client to query the Instance MetaData Service (IMDS) v2 from AWS EC2 instances.
+ * 
+ * @since 5.0
+ * @see <a href="https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html">Retrieve instance metadata</a>
+ */
+public class IMDS2Client {
+
+    private static final String ROLE_PATTERN = "[-\\w+=,.@]{1,64}";
+    private static final String IMDS_ENDPOINT = "http://169.254.169.254";
+    private static final int TOKEN_TTL_SECONDS = 21600; // 6h (default & max value)
+
+    String apiToken = null;
+
+    private IMDS2Client() {
+    }
+
+    public static IMDS2Client create() {
+        return new IMDS2Client();
+    }
+
+    /**
+     * Get the currently associated role / instance profile for this EC2 instance.
+     * 
+     * @return The name of the role (technically, instance profile) associated with
+     *         the EC2 instance from which this code is run; null if not associated.
+     */
+    public String getInstanceProfileName() {
+        String role = getMetaData("iam/security-credentials", 404);
+        if (role == null) {
+            return null;
+        }
+        if (!role.matches(ROLE_PATTERN)) {
+            throw new RuntimeException("Unable to read the instance profile name");
+        }
+        return role;
+    }
+
+    /**
+     * Query IMDSv2 to obtain credentials to access other AWS services, using the
+     * currently associated role in the instance profile.
+     * 
+     * @return Credentials in the form [accessKeyId, secretAccessKey, token].
+     */
+    public String[] getCredentials() {
+        String role = getInstanceProfileName();
+        if (role == null) {
+            throw new RuntimeException("This EC2 instance seems not to be associated to an instance profile");
+        }
+        return getCredentials(role);
+    }
+
+    /**
+     * Query IMDSv2 to obtain credentials to access other AWS services, using the
+     * specified role.
+     * 
+     * @param role The role / instance profile providing the credentials.
+     * @return Credentials in the form [accessKeyId, secretAccessKey, token].
+     */
+    public String[] getCredentials(String role) {
+        String response = getMetaData("iam/security-credentials/" + role);
+        Map<String, ?> credentials = JsonReader.jsonToMaps(response);
+        return new String[] {
+                (String) credentials.get("AccessKeyId"),
+                (String) credentials.get("SecretAccessKey"),
+                (String) credentials.get("Token")
+        };
+    }
+
+    /**
+     * Obtain a token to authorize queries to IMDSv2.
+     */
+    private String getApiToken() {
+        if (apiToken != null) // TODO: check token TTL if long-lived client
+            return apiToken;
+        try {
+            URL url = new URL(IMDS_ENDPOINT + "/latest/api/token");
+            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+            conn.setRequestMethod("PUT");
```

Comment:
A connect timeout is necessary here, something like 2 or 3 seconds should be enough

Reply:
I've added 3s timeout for the API token and 10s for the other calls.